tests/two_d_array_ut/two_d_array_ut.c

```diff
@@ -0,0 +1,642 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+
+#include <stdlib.h>
+#include <stddef.h>
+
+#include "macro_utils/macro_utils.h"
+
+#include "testrunnerswitcher.h"
+
+#include "umock_c/umock_c_negative_tests.h"
+
+#include "umock_c/umock_c.h"
+
+#define ENABLE_MOCKS
+#include "c_pal/gballoc_hl.h"
+#include "c_pal/gballoc_hl_redirect.h"
+#undef ENABLE_MOCKS
+
+#include "real_gballoc_hl.h"
+
+#include "c_pal/thandle.h"
+#include "c_util/two_d_array.h"
+
+/*TWO_D_ARRAY with regular types*/
+TWO_D_ARRAY_DEFINE_STRUCT_TYPE(uint32_t)
+THANDLE_TYPE_DECLARE(TWO_D_ARRAY_TYPEDEF_NAME(uint32_t));
+THANDLE_TYPE_DEFINE(TWO_D_ARRAY_TYPEDEF_NAME(uint32_t));
+
+TWO_D_ARRAY_TYPE_DECLARE(uint32_t);
+TWO_D_ARRAY_TYPE_DEFINE(uint32_t);
+
+typedef struct A_TEST_TAG
+{
+    int a;
+}A_TEST;
+
+THANDLE_TYPE_DECLARE(A_TEST);
+THANDLE_TYPE_DEFINE(A_TEST);
+
+TWO_D_ARRAY_DEFINE_STRUCT_TYPE(THANDLE(A_TEST))
+THANDLE_TYPE_DECLARE(TWO_D_ARRAY_TYPEDEF_NAME(THANDLE(A_TEST)));
+THANDLE_TYPE_DEFINE(TWO_D_ARRAY_TYPEDEF_NAME(THANDLE(A_TEST)));
+
+/*TWO_D_ARRAY works with THANDLEs too*/
+TWO_D_ARRAY_TYPE_DECLARE(THANDLE(A_TEST));
+TWO_D_ARRAY_TYPE_DEFINE(THANDLE(A_TEST));
+
+MU_DEFINE_ENUM_STRINGS(UMOCK_C_ERROR_CODE, UMOCK_C_ERROR_CODE_VALUES)
+
+static void on_umock_c_error(UMOCK_C_ERROR_CODE error_code)
+{
+    ASSERT_FAIL("umock_c reported error :%" PRI_MU_ENUM "", MU_ENUM_VALUE(UMOCK_C_ERROR_CODE, error_code));
+}
+
+
+BEGIN_TEST_SUITE(TEST_SUITE_NAME_FROM_CMAKE)
+
+TEST_SUITE_INITIALIZE(it_does_something)
+{
+    ASSERT_ARE_EQUAL(int, 0, real_gballoc_hl_init(NULL, NULL));
+
+    umock_c_init(on_umock_c_error);
+
+    REGISTER_GBALLOC_HL_GLOBAL_MOCK_HOOK();
+}
+
+TEST_SUITE_CLEANUP(TestClassCleanup)
+{
+    umock_c_deinit();
+
+    real_gballoc_hl_deinit();
+}
+
+TEST_FUNCTION_INITIALIZE(f)
+{
+    umock_c_negative_tests_init();
+    umock_c_reset_all_calls();
+}
+
+TEST_FUNCTION_CLEANUP(cleans)
+{
+    umock_c_negative_tests_deinit();
+}
+
+/*TWO_D_ARRAY_CREATE(T)*/
+
+/* Tests_SRS_TWO_D_ARRAY_07_003: [ TWO_D_ARRAY_CREATE(T) shall call THANDLE_MALLOC_FLEX with TWO_D_ARRAY_FREE(T) as dispose function, nmemb set to row_size and size set to sizeof(T*). ]*/
+/* Tests_SRS_TWO_D_ARRAY_07_004: [ TWO_D_ARRAY_CREATE(T) shall set all rows point to NULL. ]*/
+/* Tests_SRS_TWO_D_ARRAY_07_006: [ TWO_D_ARRAY_CREATE(T) shall succeed and return a non-NULL value. ]*/
+TEST_FUNCTION(TWO_D_ARRAY_CREATE_create_uint32_t_succeeds)
+{
+    //arrange
+    STRICT_EXPECTED_CALL(malloc_flex(IGNORED_ARG, 5, sizeof(uint32_t*)));
+
+    //act
+    TWO_D_ARRAY(uint32_t) tdarr = TWO_D_ARRAY_CREATE(uint32_t)(5, 5);
```

Comment:
>5, 5 [](http://example.com/codeflow?start=63&length=4)

We usually test at the boundaries (inflexion points in the range of values).

Imagine that the range of values for rows is 0 .. UINT32_MAX.
These inflexion points are the points where we expect behavior to be different.

For example I would test what happens when called with 0 - we'd expact that to fail, right?
Also we should test that when called with 1 for rows a 2D array is allocated.

from 1 to UINT32_MAX-1 the expected behavior is the same, so it adds very little value to test for 2 and 3 and 4. So a best practice is to pick something in the interval and test to see the behavior is kept (5 is Ok to test with for example).

So, we should have tests for row values:
0, 1, 5, UINT32_MAX - 1 and UINT32_MAX).

Reply:
I guess this would be UINT64_MAX, so you won't actually be able to allocate that much memory, but you can trick the test to not allocate anything, but rather just return a dummy pointer.

Reply:
The beauty of hooks

Reply:
Thank you so much for explaining the details. I tried to test UINT64_MAX before and that much memory can't be allocated. I try to return a dummy pointer now, but in the two_d_array_ll.h line 125-128, it try to access the returned address by malloc and set every row to NULL, which will give a access violation when I set return a dummy pointer in malloc_flex. I'm confused about whether UINT64_MAX could be tested here.

Reply:
Yes, there is no computer with that kind of memory.
For mocking, yes I was thinking that you can return a dummy value, but looks like because the constructor is initializing the memory that won't be doable (without extracting the row access and more mocking it, which would be rather unnecessary).

So we won't write any test with UINT64_MAX I assume.
How about a test with UINT32_MAX + 1, do we test that?

Reply:
Do we really need uint64_t? I don't think we expect having 4B rows or 4B columns.
Frankly uint32_t should be enough IMO. Then you can also test the extremeties of the interval.

Reply:
done